### ISSUE TITLE
Perf: Native Vs Managed performance differences

### ISSUE THREAD

**Issue by Wraith2** [dotnet/SqlClient#375]
I did a little investigation on the difference between native and managed performance. I did this by switching two things, 1) the network layer, 2) the pool implementation. The https://github.com/aspnet/DataAccessPerformance repo has a project which measures the throughput of sqlclient and this was run on my desktop with various release dev builds of the current master , async mode 32 threads for 60 seconds and 3 second warmup to get everything tiered and connected. All tests are on netcore3.1 and windows. The results are:

Throughput | Network | Pool 
 ------------- | -- | -- 
|34427|managed | managed
54443|managed | native
36441|native | managed
63804|native | native

The current library uses native mode on win32 windows and managed mode in any other environment (including UAP on windows). From the table above you can see that the major difference in performance is when using the managed pool implementation. but because the native and managed network and pool are controlled by the same setting you can't tell them apart easily unless you play with them like this.

The important part of the pools that causes the effect is creating a pool key to look up in the pool concurrent dictionary. 

* In native mode the windows thread context is used to fetch the current user sid and that is used as part of the key, 
https://github.com/dotnet/SqlClient/blob/master/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/ProviderBase/DbConnectionPoolIdentity.Windows.cs . This behaviour is actually pretty lightweight and will perform correctly in the presence of impersonation. 

* In managed mode the current domain name and user name are fetched from the runtime Environment properties that expose them and then stuck together, https://github.com/dotnet/SqlClient/blob/master/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/ProviderBase/DbConnectionPoolIdentity.cs This is probably correct when impersonating because the environment property implementations don't use environment variables they pinvoke into the platform which will take notice of impersonation and do the right thing. On non-windows platforms where impersonation isn't possible it wouldn't make a difference I don't think. On windows the pinvokes required to do this take significant time and are slightly wasteful of resources.

Currently the only way to get managed network implementation for users is to be running on UAP or a non-windows platform. They are forced to take the pool implementation that matches the network implementation, it has no configurability. I'd like to discuss whether it's a good idea to split these two things and possibly choose the best choice for them in a finer grained way.

Even with the native pool it's clear that the managed implementation has room for improvement before it's ready to be the default for everyone but the difference isn't as bad as i'd thought and with the native pool being an option it would be easier to see the hotspots in the network layer.

**Comment by cheenamalhotra** [comment]
@Wraith2 

Do you have a proposition for the same on how this can be achieved?
I'll let @David-Engel take a call on the same, from product shipping perspective.

**Comment by Wraith2** [comment]
I have no concrete suggestions just general ideas which is why I thought I'd post numbers and show what I've found. I think it would be interesting to get some windows vs Linux numbers from a single system in managed mode to see if the managed strategy is just poor on windows. 

I'd like to make the choices of pool and network strategy switchable separately in debug and release builds if possible which isn't a big change, it's a policy decision.

Long term I think removing the need for the native sni.dll is a good idea. It's closed source which is a bit of a licensing issue, it bottlenecks any issues that trace into it (I haven't seen any yet but I'm sure it gets fixes) and it causes loader problems. Getting the managed network layer up to the same speed and removing the need for the native shim would reduce complexity and allow the network layer interface to be evolved for better performance and maintainability.

Even longer term I'd thought a new network layer build on pipelines as used in asp.net for high performance scenarios might simplify the current managed version and offer more speed improvements.

Overall I'm not angling for any particular plan of action but having worked on the code for a while I have ideas about things that could be done to improve developer and customer experience with it. I'm not going to run off and re-work fundamental systems without knowing whether it's a plausible avenue to spend time on.

**Comment by David-Engel** [comment]
@Wraith2 
I would personally love to get rid of the native layer. Thanks for peeling back the covers a bit and exposing where some of the performance differences lie. The only reason I'm aware of to keep the native layer on Windows is performance. There was some functional difference at one point, but I think that is gone since we moved to MSAL for AAD auth.

cc: @saurabh500

**Comment by Wraith2** [comment]
I did some further comparisons. This time I measured netfx vs core

Configuration | Throughput | Scaled %
-- | -- | --
netcore 3.1 native | 66416 | 100
netfx462 native | 49292 | 74.21706
netcore3.1 managed | 56154 | 84.5489

Even in managed mode where it has it's worst performance the netcore managed build can get higher throughput than the netfx build can on native. Native on netcore is can do 30% more work than equivalent on netfx. Some of that will be the runtime, some the library.

Anecdotally I think the netcore build could go faster and that i'm reaching cpu limits of my hardware. When running it my fans spin up and the cpu usage is much more heavily weighted towards the sql server process than with the other builds, i could run get higher TPS from the library if I could feed it faster.
